<launch>
  
  <arg name="UAV_NAME" />
  <arg name="rosbag/topic/points" />
  <arg name="rosbag/directory" />
  <arg name="rosbag/filename" />

  <!-- ENV VARS -->
  <arg name="ENABLE_PROFILER" default="false" />

  <arg name="standalone" default="true" />
  <arg name="debug" default="false" />
  
  <!-- CUSTOM CONFIGS -->
  <arg name="config/feature_extraction" default="" />
  <arg name="config/feature_selection" default="" />
  <arg name="config/aloam" default="" />
  <arg name="config/aloam/rosbag" default="" />
  <arg name="config/aloam/sensor" default="" />

  <!-- DEFAULT PARAMETERS AND FRAMES -->
  <arg name="fcu_frame" default="$(arg UAV_NAME)/fcu" />
  <arg name="lidar_frame" default="$(arg UAV_NAME)/os_sensor" />
  <arg name="odom_frame" default="$(arg UAV_NAME)/slam_odom_origin" />
  <arg name="map_frame" default="$(arg UAV_NAME)/slam_origin" />

  <arg     if="$(arg debug)" name="launch_prefix" value="debug_roslaunch" />
  <arg unless="$(arg debug)" name="launch_prefix" value="" />

  <arg name="launch_delay" default="0" />
  <arg name="nodelet_manager_name" default="" />
  <arg     if="$(eval arg('standalone') or arg('debug'))" name="nodelet" value="standalone" />
  <arg unless="$(eval arg('standalone') or arg('debug'))" name="nodelet" value="load" />
  <arg     if="$(eval arg('standalone') or arg('debug'))" name="nodelet_manager" value="" />
  <arg unless="$(eval arg('standalone') or arg('debug'))" name="nodelet_manager" value="$(arg nodelet_manager_name)" />

  <group ns="$(arg UAV_NAME)">

    <node pkg="nodelet" type="nodelet" name="aloam" args="$(arg nodelet) aloam_slam/AloamSlamRosbag $(arg nodelet_manager)" output="screen" launch-prefix="bash -c 'sleep $(arg launch_delay); $0 $@'; $(arg launch_prefix)">

      <rosparam file="$(find feature_extraction)/config/default.yaml" command="load" />
      <rosparam file="$(find feature_selection)/config/default.yaml" command="load" />
      <rosparam file="$(find aloam_slam)/config/aloam.yaml" command="load" />
      <rosparam file="$(find aloam_slam)/config/aloam_rosbag.yaml" command="load" />
      <rosparam file="$(find aloam_slam)/config/sensor.yaml" command="load" />

      <rosparam if="$(eval arg('config/feature_extraction') != '')" file="$(arg config/feature_extraction)" command="load" />
      <rosparam if="$(eval arg('config/feature_selection') != '')" file="$(arg config/feature_selection)" command="load" />
      <rosparam if="$(eval arg('config/aloam') != '')" file="$(arg config/aloam)" command="load" />
      <rosparam if="$(eval arg('config/aloam/rosbag') != '')" file="$(arg config/aloam/rosbag)" command="load" />
      <rosparam if="$(eval arg('config/aloam/sensor') != '')" file="$(arg config/aloam/sensor)" command="load" />

      <param name="enable_profiler" type="bool" value="$(arg ENABLE_PROFILER)" />

      <param name="uav_name" type="string" value="$(arg UAV_NAME)" />
      <param name="fcu_frame" type="string" value="$(arg fcu_frame)" />
      <param name="odom_frame" type="string" value="$(arg odom_frame)" />
      <param name="map_frame" type="string" value="$(arg map_frame)" />
      <param name="lidar_frame" type="string" value="$(arg lidar_frame)" />

      <param name="rosbag/points_topic" type="string" value="$(arg rosbag/topic/points)" />
      <param name="rosbag/directory" type="string" value="$(arg rosbag/directory)" />
      <param name="rosbag/filename" type="string" value="$(arg rosbag/filename)" />

      <!-- Publishers -->
      <remap from="~diag_out"        to="slam/diagnostics"/>
      <remap from="~path_out"        to="slam/path"/>
      <remap from="~map_out"         to="slam/map"/>
      <remap from="~eigenvalues_out" to="slam/eigenvalues"/>

      <remap from="~odom_local_out"  to="slam/odom/local" />
      <remap from="~odom_global_out" to="slam/odom" />

      <remap from="~points_out"           to="slam/points" />
      <remap from="~edges_out"            to="slam/features/edges/extracted" />
      <remap from="~planes_out"           to="slam/features/planes/extracted" />
      <remap from="~edges_salient_out"    to="slam/features/edges/salient" />
      <remap from="~planes_salient_out"   to="slam/features/planes/salient" />
      <remap from="~edges_selected_out"   to="slam/features/edges/selected" />
      <remap from="~planes_selected_out"  to="slam/features/planes/selected" />
      <remap from="~signatures_out"       to="slam/features/signatures" />
      <remap from="~cloud_range_diff_out" to="slam/cloud/range_diff" />

    </node>

  </group>

</launch>
